sdk/core/azure-core-test/inc/azure/core/test/test_base.hpp

```diff
@@ -262,8 +280,11 @@ namespace Azure { namespace Core { namespace Test {
      * @brief Run before each test.
      *
      */
-    void SetUpTestBase(std::string const& baseRecordingPath)
+    void SetUpTestBase(std::string const& serviceName, std::string const& baseRecordingPath)
```

Comment:
Since Attestation is the only one who needs this and won't use the test-resourses.json to set its env vars, can you move this new arg to the end as optional with default value empty. Then, only Attestation tests would use it.

We can also rename the input arg to something like `envVarPrefix` and add a note about what it does for you when it is  passed to SetUpTestBase.

This way we don't need to add the changes to Key Vault and Storage tests. Since they don't need it

Reply:
Attestation *does* use test-resources.json to set its environment variables, just not these three. That's because it's *important* that test-resources.json files be able to be shared across azure SDKs (similar to the test-resources-pre.ps1 scripts).

I will be confirming that KeyVault and STorage's live tests continue to work - this change should *not* require any change to them except for adding the words "KEYVAULT" and "STORAGE" to the ctor.

All the other SDKs call that variable "directory" - I thought "serviceName" was more descriptive. But I could call it servicePrefix or something like that.

Reply:
I profoundly disagree with this assertion. Somehow I (and others) have managed to develop SDKs for the 4 primary azure languages without modifying the test-resources.json. But for some reason, C++ is different from all the other languages.

I'm still trying to understand why the same solution that every other language has implemented is not idiomatic C++.  Could you please help me understand which C++ idioms this breaks?

As far as I can tell, my change simply updates the C++ SDK to use the same patterns that every other Azure SDK does - it updates the SetUpTestBase method to include the serviceDirectory, exactly as is done for all the other languages (each language has an idiomaticly appropriate version for the language - for Python, it's done in the `EnvironmentVariableLoader`, for C#, it's handled in the `TestEnvironment` class. Sometimes it's not obvious where it is, but a similar mapping exists for all the Azure languages I've worked with.

Reply:
Well, we are introducing a business-rule to our code based for `keeping a test cofing file` portable across languages. In C++, the idiomatic rules for the code (the things why people would choose C++ and not other languages) are things like performance, memory, or something related.  It's not expected to see a pattern where API needs an input value so it can make test-configuration (and only the test-configuration, not the test-cases or test-data) portable to other languages.

If we would really like to make the test-configuration the same for all languages, it would need to be in the eng/common folder. Or pulled from some repository.  Whatever that is not a duplicated copy for each language.

If in the future, how do other language's developer know about this contract?  How do we avoid breaking it?  What if there's a features that go live only in one language in the future and requires changes to the test-configuration. There are multiple things why I think keeping that file portable is not easy to maintain and does not bring too much value against the effort to keep it portable.

Reply:
This issue *only* applies to service implementers. It does not apply to the developers who use the SDK. Developers who use the SDK always use AZURE_CLIENT_ID, AZURE_TENANT_ID, and AZURE_CLIENT_SECRET. 

Service implementers follow the patterns established by the SDK - so for Python developers, they pass the serviceDirectory parameter from new-testresources.ps1 to the PowerShellPreparer (now called EnvironmentVariableLoader) which is used for their tests. For C#, the sdk figures out the serviceDirectory value by looking at the running assembly (it also guesses the recordings directory from the assembly), other languages have similar checks - some are explicit, some are implicit.

What I'm proposing in this pull request is that the pattern for C++ service implementers should include them specifying the serviceDirectory parameter that they specified in their call to new-testresources.ps1.

Also, the testresources.json defines the service specific configuration parameters, and the test-resources-pre.ps1 and test-resources-post.ps1 scripts perform actions before and after the testresources ARM template for the service is run. They do *not* belong in the eng\common directory, because they're not common to all services. Instead they belong in a per-service directory. And the should be able to be shared between different languages because fundamentally each language is interacting with the same service.

If a service brings up new functionality that requires that the arm template for the service be modified (infrequent, but it happens), the arm template should be able to be shared across different languages without modification - that way as each service's SDK team implements the new functionality, they operate across a common service baseline.

Reply:
> This issue only applies to service implementers. It does not apply to the developers 

Everyone from the community can be a service implementer.

> Service implementers follow the patterns established by the SDK

We don't have one for C++.  Are we trying to follow what other language's decided instead of creating the pattern for C++ within the team?  

> If a service brings up new functionality that requires that the arm template for the service be modified (infrequent, but it happens), the arm template should be able to be shared across different languages without modification 

Can you update the attestation arm template and add the env vars mapping in the outputs like the other services in C++?  And then share the updated arm template across languages?  It should still work for all other languages.

Reply:
> Can you update the attestation arm template and add the env vars mapping in the outputs like the other services in C++? And then share the updated arm template across languages? It should still work for all other languages.

This is the crux of my issue.  Every language other than C++ solves the problem of mapping <serviceDirectory>_XXX to AZURE_XXX inside the test framework.

C++ is special, requiring modifications to the test-resources.json for live tests to pass, and those modifications are inappropriate for the patterns used in all the other Azure languages I'm familiar with.

I'm pushing HARD against that pattern, because IMHO test-resources.json files should not be language specific.

FWIW, my current version of the solution solves this problem entirely inside the test framework, without requiring any modification to tests or json files.

Reply:
@weshaggard  @benbp @danieljurek  Is C++ violating any rule by using test-resources.json to do `env var mapping` as prt of the `outputs`? like:

https://github.com/Azure/azure-sdk-for-cpp/blob/main/sdk/storage/test-resources.json#L323

If yes, we might want to update and fix this for all C++ services, so we keep the pattern consistent within C++